### PR TITLE
[ECO-4687]Feature/resume recover

### LINKED ISSUE #393: [ably-ruby][no-connection-serial] Implement connection resume/recover
- Implement connection resume/recover as per no-connection-serial spec
- related to #378

### PR DESCRIPTION
**Pull request by sacOO7**
- Fixed #393 
- There might be failing tests on the PR. Those will be addressed as a part of https://github.com/ably/ably-ruby/issues/392 and https://github.com/ably/ably-ruby/issues/378

### CODE REVIEW

**Comment by sacOO7** [comment]
There might be failing tests on the PR. This will be fixed as a part of separate issue created https://github.com/ably/ably-ruby/issues/378

**Review by lawrence-forooghian — COMMENTED** [review]
@sacOO7 I am not sure how to review this PR. There's no overall description to explain in detail what changes it aims to make (e.g. which spec points it aims to implement).

So I thought I’d try reviewing it commit-by-commit, hoping that the commit messages would give me useful information. But immediately I found that the commits haven't been structured with reviewability in mind. For example, 1d0f95931031f213587b6456d808c0d24814bd01 implements a class without any tests, and then the next commit (07c053fc23864cf7b9b2c783c471ec3f22d9bcb6) _fixes some code from the previous commit_ (i.e. meaning that someone reviewing commit-by-commit has to look at the initial attempt, spend some time getting confused by it because it's wrong, and then eventually realise that it gets fixed in the next commit), and they spend time getting confused about why the commit that introduces the class has no tests, only to again realise that they get introduced in the next commit, and they spend time wondering why there is no test for what happens when the recovery key can't be decoded, only to eventually realise that for some reason this test is added about 10 commits later (d654665604fb6f60c0a0b00d3e130a67ad925e35).

And I’d argue that given that this PR contains a mixture of features and refactors and typo fixes, it _needs_ to be reviewable commit-by-commit; there’s too much noise for the PR to be reviewable simply by looking at the overall diff.

Do you have any suggestions for how I can review this PR in its current state? If not, please could you put it back into draft and try to make it more reviewable?

**Comment by lawrence-forooghian** [comment]
I was wondering if you'd seen https://github.com/ably/engineering/blob/33da110db02504e329109fd4618eee76d79dd0ec/best-practices/commits.md, which explains some best practice for Git commit / history structure? I think that some of the guidance there might be helpful here.

**Comment by sacOO7** [comment]
> > There might be failing tests on the PR. Those will be addressed as a part of (…)
> 
> We can't merge a PR with failing tests into `main` since it will stop us from being able to do a release. I suggest that you create an integration branch e.g. `integration/protocol-v2` and make this PR target that branch (once #391 is merged).

Yes, I have already created a branch `feature/no-connection-serial`. `fix/tests` will be merged into the same.

**Comment by lawrence-forooghian** [comment]
> About fix/tests, I was planning to merge it into feature/no-connection-serial, but we can also merge it into main once all review comments are addressed.

Yeah, I agree that that one should go into `main` without waiting for all of the no connection serial work to be merged.

**Comment by sacOO7** [comment]
> @sacOO7 I am not sure how to review this PR. There's no overall description to explain in detail what changes it aims to make (e.g. which spec points it aims to implement).
> 
> So I thought I’d try reviewing it commit-by-commit, hoping that the commit messages would give me useful information. But immediately I found that the commits haven't been structured with reviewability in mind. For example, [1d0f959](https://github.com/ably/ably-ruby/commit/1d0f95931031f213587b6456d808c0d24814bd01) implements a class without any tests, and then the next commit ([07c053f](https://github.com/ably/ably-ruby/commit/07c053fc23864cf7b9b2c783c471ec3f22d9bcb6)) _fixes some code from the previous commit_ (i.e. meaning that someone reviewing commit-by-commit has to look at the initial attempt, spend some time getting confused by it because it's wrong, and then eventually realise that it gets fixed in the next commit), and they spend time getting confused about why the commit that introduces the class has no tests, only to again realise that they get introduced in the next commit, and they spend time wondering why there is no test for what happens when the recovery key can't be decoded, only to eventually realise that for some reason this test is added about 10 commits later ([d654665](https://github.com/ably/ably-ruby/commit/d654665604fb6f60c0a0b00d3e130a67ad925e35)).
> 
> And I’d argue that given that this PR contains a mixture of features and refactors and typo fixes, it _needs_ to be reviewable commit-by-commit; there’s too much noise for the PR to be reviewable simply by looking at the overall diff.
> 
> Do you have any suggestions for how I can review this PR in its current state? If not, please could you put it back into draft and try to make it more reviewable?

I agree we don't have atomic commits. This partially also has to do with a fact that current implementation is split across codebase for a given atomic spec ( also pointed out in the comment -> https://github.com/ably/ably-ruby/issues/407#issuecomment-2129397204 ). 
Maybe I will try to add more description for changes being implemented at the top of the PR.

**Comment by lawrence-forooghian** [comment]
> This partially also has to do with a fact that current implementation is split across codebase for a given atomic spec

That has nothing to do with the example I gave above though (a feature which is implemented in a broken fashion without tests, then some tests are added and the code is fixed, then later on another test that should have been there in the first place is added). That's just a matter of making sure that once you've finished your work, you take some time to fix the commit history.

**Comment by lawrence-forooghian** [comment]
> I will try to add more description for changes being implemented at the top of the PR. Also, will try to add more meaningful commits going forward. For now, I would request to the review the PR code changes.

I will attempt to review this PR. But please think about reviewability in your future PRs. It would make my life as a reviewer a lot easier.

**Comment by lawrence-forooghian** [comment]
> Also, at the end, we will have a final working PR against main with full no-connection-serial implementation, so the code itself will make sense and there won't be a need to look at the commits.

Sure, but I still need to be able to review the PRs.

**Comment by lawrence-forooghian** [comment]
And, as stated in the linked commit guidance, "the history is the deliverable, not the code".

File: lib/ably/realtime/client.rb [hunk 1]
```diff
@@ -62,7 +62,7 @@ class Client
 
       # When a recover option is specified a connection inherits the state of a previous connection that may have existed under a different instance of the Realtime library, please refer to the API documentation for further information on connection state recovery
       # @return [String,Nil]
-      attr_reader :recover
+      attr_accessor :recover
```
Comment (lawrence-forooghian) [inline lib/ably/realtime/client.rb:65]: Why was this changed?

File: lib/ably/realtime/connection.rb [hunk 2]
```diff
@@ -347,41 +333,48 @@ def internet_up?
...
+      # The recovery key string can be used by another client to recover this connection's state in the recover client
+      # options property. See connection state recover options for more information.
+      #
+      # @spec RTN16g
+      #
+      # @return [String] a json string which incorporates the @connectionKey@, the current @msgSerial@ and collection
+      # of pairs of channel @name@ and current @channelSerial@ for every currently attached channel
+      def create_recovery_key
+        if key.nil? || key.empty? || state == :closing || state == :closed || state == :failed || state == :suspended
+          return "" #RTN16h
```
Comment (lawrence-forooghian) [inline lib/ably/realtime/connection.rb:358]: this spec point doesn't exist?

**Review by lawrence-forooghian — COMMENTED** [review]
I had a cursory look at this PR. A few comments:

- if you're going to split the work across multiple PRs, then I think you're going to have to be the one responsible for making sure that all of the required spec points end up being implemented, because I don't know what the scope of this PR is meant to be
- there seems to be new code added, does this new code not need tests?

File: lib/ably/realtime/client.rb (hunk 1 above)
Comment (sacOO7) [reply lib/ably/realtime/client.rb:65]: updated at #409

**Comment by sacOO7** [comment]
@lawrence-forooghian I have created a separate PR with proper commit history -> #409 
I have tried to keep changes atomic and code easy to understand. I will also add PR summary, so it will be easy to review the code. That said, I will close this PR and discussion can be continued on #409
